_posts/2019-03-30-numba-stencil.md

```diff
@@ -0,0 +1,323 @@
+3.  Dask Array for parallelizing array computations across many chunks
+4.  NumPy's Generalized Universal Functions (gufuncs) to tie everything
+    together
+
+In the end we'll show how a novice developer can write a small amount of Python
+to efficiently compute localized computation on large amounts of data.  In
+particular we'll write a simple function to smooth images and apply that in
+parallel across a large stack of images.
+
+Here is the full code, we'll dive into it piece by piece below.
+
+```python
+import numba
+
+@numba.stencil
+def _smooth(x):
+    return (x[-1, -1] + x[-1, 0] + x[-1, 1] +
+            x[ 0, -1] + x[ 0, 0] + x[ 0, 1] +
+            x[ 1, -1] + x[ 1, 0] + x[ 1, 1]) // 9
+
+
+@numba.guvectorize(
+    [(numba.int8[:, :], numba.int8[:, :])],
+    '(n, m) -> (n, m)'
+)
+def smooth(x, out):
+    out[:] = _smooth(x)
+
+
+# If you want fake data
+import dask.array as da
+x = da.ones((1000000, 1000, 1000), chunks=('auto', -1, -1), dtype='int8')
+
+# If you have actual data
+import dask_image
+x = dask_image.imread.imread('/path/to/*.png')
+
+y = smooth(x)
+# dask.array<transpose, shape=(1000000, 1000, 1000), dtype=int8, chunksize=(125, 1000, 1000)>
+```
+
+Lets break this down a bit
+
+
+Numba Stencils
+--------------
+
+**Docs:**: https://numba.pydata.org/numba-doc/dev/user/stencil.html
+
+Many array computing functions operate only on a local region of the array.
+This is common in image processing, signals processing, simulation, the
+solution of differential equations, anomaly detection, time series analysis,
+and more.  Typically we write code that looks like the following:
+
+```python
+def _smooth(x):
+    out = np.empty_like(x)
+    for i in range(1, x.shape[0] - 1):
+        for j in range(1, x.shape[1] - 1):
+            out[i, j] = x[i + -1, j + -1] + x[i + -1, j + 0] + x[i + -1, j + 1] +
+                        x[i +  0, j + -1] + x[i +  0, j + 0] + x[i +  0, j + 1] +
+                        x[i +  1, j + -1] + x[i +  1, j + 0] + x[i +  1, j + 1]) // 9
+
+    return out
+```
+
+Or something similar.  The `numba.stencil` decorator makes this a bit easier to
+write down.  You just write down what happens on every element, and Numba
+handles the rest.
+
+```python
+@numba.stencil
+def _smooth(x):
+    return (x[-1, -1] + x[-1, 0] + x[-1, 1] +
+            x[ 0, -1] + x[ 0, 0] + x[ 0, 1] +
+            x[ 1, -1] + x[ 1, 0] + x[ 1, 1]) // 9
+```
+
+
+Numba JIT
+---------
+
+**Docs:** http://numba.pydata.org/
+
+When we run this function on a NumPy array, we find that it is slow, operating
+at Python speeds.
+
+```python
+x = np.ones((100, 100))
+timeit f(x)
+527 ms ± 44.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+```
+
+But if JIT compile this function with Numba, then it runs more quickly.
+
+```python
+@numba.njit
+def smooth(x):
+    return _smooth(x)
+
+%timeit smooth(x)
+70.8 µs ± 6.38 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
+```
+
+For those counting, that's over 1000x faster.
```

Comment:
Might be worth comparing to equivalent functions from SciPy ndimage and/or scikit-image.

Reply:
Do you happen to know the operation to do here?

Reply:
I think there's no direct operation for a mean filter, but you can use `scipy.ndimage.filters.convolve`, with `weights=[[1/9.0]*3]*3`.

Reply:
Actually, you could also use `scipy.ndimage.uniform_filter` with `size=3` here.